### DISCUSSION TITLE
How to provide `executableArgs` to debuggable

### DISCUSSION THREAD

**Discussion by gollth** [mrcjkb/rustaceanvim#163] [Q&A] [answered]
Hi `rustaceanvim`-team,

first of all thanks for this rework, it is working great for me. One thing I am struggling with, though, is to pass custom CLI arguments to the executable target I want to debug. With `vim.cmd.RustLsp("debuggables")` this only shows a list of available targets. From a quick glance at the code I saw there is [`args.executableArgs`](https://github.com/mrcjkb/rustaceanvim/blob/master/lua/rustaceanvim/dap.lua#L333) field, but I don't know how I can control that from my config. 

Any ideas? =)

**Comment by mrcjkb** [comment]
Hey :wave:

`executableArgs` is a field in rust-analyzer's response to the `runnables` LSP request.
There's no way to influence that.

What you  can do is create a DAP launch configuration (see [`:h dap-launch.json`](https://github.com/mfussenegger/nvim-dap/blob/9adbfdca13afbe646d09a8d7a86d5d031fb9c5a5/doc/dap.txt#L316)).
This plugin automatically calls `require('dap.ext.vscode').load_launchjs()` and uses any `.vscode/launch.json` that nvim-dap finds as a base for the debuggables configurations.

[Here's an example](https://gist.github.com/xanathar/c7c83e6d53b72dd4464f695607012629). You can pass in CLI arguments via `args`.

**Reply by gollth** [reply]
Hi thanks for the quick answer. Make sense to me to use this launch configuration file. I'm struggeling a bit with the proper path, though. I have the following config in `<repo-root>/.vscode/launch.json`:

```json
{
  "version": "0.2.0",
  "configurations": [{
    "name": "Launch an application",
    "type": "lldb",
    "request": "launch",
    "cargo": {
      "args": ["run", "--bin", "my_binary"]
    },
    "args": [ "--my-cli-arg" ],
    "cwd": "${workspaceRoot}/subfolder"
  }]
}
```

When I now start the debugger via `vim.cmd.RustLsp("debuggables")` the config does not get picked up it seems. Printing [resolved_path in nvim-dap](https://github.com/mfussenegger/nvim-dap/blob/9adbfdca13afbe646d09a8d7a86d5d031fb9c5a5/lua/dap/ext/vscode.lua#L181) shows the correct path, though. Any ideas what I am missing?

**Reply by mrcjkb** [reply]
`RustLsp debuggables` will only use the launch configuration as a base to be populated with the information it gets from rust-analyzer.

I believe if you want to use your configuration directly, you should run `:lua  instead.

🤔 Maybe it's a good idea to extend `:RustLsp debuggables` to also list predefined launch configurations. I'll have to think about it.

By the way, since 3.16.0, this plugin tries to add the debuggables from rust-analyzer to the nvim-dap configurations table on startup.
So you should be able to just use `:lua require('dap').continue()` instead of `:RustLsp debuggables`.

**Reply by gollth** [reply]
I think the command got lost in the formatting in your previous post. I assumed you meant sth. like `:lua require('dap.ext.vscode').load_launchjs()`, but this doesn't seem to work either. For my use case, I hacked now the `rustaceanvim/dap.lua` locally. Not pretty, but does the job: 

```diff
-      final_config.args = args.executableArgs or {}
+      final_config.args = vim.fn.split(vim.fn.input { prompt = 'Arguments: ' })
```

**Reply by gollth** [reply]
> Maybe it's a good idea to extend :RustLsp debuggables to also list predefined launch configurations.

That would be super convenient ^^

**Reply by mrcjkb** [reply, marked as answer]
I've just released version `3.16.1` of the plugin, which [fixes loading launch.json configs](https://github.com/mrcjkb/rustaceanvim/pull/167).

After some thought, I have decided it is not a good idea for `:RustLsp debuggables` to also list predefined launch configurations.
Instead, the plugin calls `require('dap.ext.vscode').load_launchjs(nil, { lldb = { 'rust' }, codelldb = { 'rust' } })` on LSP attach, adding any `lldb` or `codelldb` configurations + the rust-analyzer configurations to nvim-dap's rust configurations.

So if you call `:lua require('dap').continue()`, it should prompt you with the configurations in your `launch.json` + the rust-analyzer ones.

**Comment by gollth** [comment]
I checked out 3.16.1 and tried to call now `require('dap').continue()` directly with my keybinding. I get prompted for args (with the default value of `--bin my_binary --`). When I accept these the plugin crashes:

```
vim/fs.lua:0: invalid value (nil) at index 1 in table for 'concat'
stack traceback:
	[C]: in function 'concat'
	vim/fs.lua: in function 'joinpath'
	...al/share/nvim/lazy/rustaceanvim/lua/rustaceanvim/dap.lua:190: in function 'on_exit'
	/usr/share/nvim/runtime/lua/vim/_system.lua:297: in function </usr/share/nvim/runtime/lua/vim/_system.lua:267>
	[C]: in function 'nvim_buf_delete'
	...h/.local/share/nvim/lazy/nui.nvim/lua/nui/popup/init.lua:296: in function '_buf_destory'
	...h/.local/share/nvim/lazy/nui.nvim/lua/nui/popup/init.lua:320: in function 'unmount'
	....local/share/nvim/lazy/noice.nvim/lua/noice/view/nui.lua:196: in function <....local/share/nvim/lazy/noice.nvim/lua/noice/view/nui.lua:193>
	[C]: in function 'xpcall'
	...local/share/nvim/lazy/noice.nvim/lua/noice/util/call.lua:149: in function <...local/share/nvim/lazy/noice.nvim/lua/noice/util/call.lua:134>
	....local/share/nvim/lazy/noice.nvim/lua/noice/view/nui.lua:209: in function 'hide'
	...local/share/nvim/lazy/noice.nvim/lua/noice/view/init.lua:145: in function 'display'
	.../share/nvim/lazy/noice.nvim/lua/noice/message/router.lua:215: in function <.../share/nvim/lazy/noice.nvim/lua/noice/message/router.lua:155>
	[C]: in function 'xpcall'
	...local/share/nvim/lazy/noice.nvim/lua/noice/util/call.lua:149: in function <...local/share/nvim/lazy/noice.nvim/lua/noice/util/call.lua:134>
	[C]: in function 'pcall'
	...local/share/nvim/lazy/noice.nvim/lua/noice/util/init.lua:120: in function ''
	vim/_editor.lua: in function ''
	vim/_editor.lua: in function <vim/_editor.lua:0>
```

I suspected that it might be my `launch.json` file, but VS Code is able to run the debug configuration (including the arguments) just fine. What am I missing here?

**Reply by gollth** [reply]
Okay, turns out I had local changes in my `nvim-dap/` file during testing 🤦. Discarding those and using absolute paths in the `launch.json` it works perfectly! Thank your🙏
